compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -4263,6 +4281,160 @@ public void visit(BLangStringTemplateLiteral stringTemplateLiteral) {
         result = rewriteExpr(constructStringTemplateConcatExpression(stringTemplateLiteral.exprs));
     }
 
+    /**
+     * The raw template literal gets desugared to a type init expression. For each literal, a new object class type
+     * def is generated from the object type. The type init expression creates an instance of this generated object
+     * type. For example, consider the following statements:
+     *      string name = "Pubudu";
+     *      'object:RawTemplate rt = `Hello ${name}!`;
+     *
+     * The raw template literal above is desugared to:
+     *      type RawTemplate$Impl$0 object {
+     *          public string[] strings = ["Hello ", "!"];
+     *          public (any|error)[] insertions;
+     *
+     *          function init((any|error)[] insertions) {
+     *              self.insertions = insertions;
+     *          }
+     *      };
+     *
+     *      // somewhere in code
+     *      'object:RawTemplate rt = new RawTemplate$Impl$0([name]);
+     *
+     * @param rawTemplateLiteral The raw template literal to be desugared.
+     */
+    @Override
+    public void visit(BLangRawTemplateLiteral rawTemplateLiteral) {
+        DiagnosticPos pos = rawTemplateLiteral.pos;
+        BObjectType objType = (BObjectType) rawTemplateLiteral.type;
+        BLangTypeDefinition objClassDef = desugarTemplateLiteralObjectTypedef(rawTemplateLiteral.strings, objType, pos);
+        BObjectType classObjType = (BObjectType) objClassDef.type;
+
+        BVarSymbol insertionsSym = classObjType.fields.get("insertions").symbol;
+        BLangListConstructorExpr insertionsList = ASTBuilderUtil.createEmptyArrayLiteral(pos, insertionsSym.type);
+        insertionsList.exprs.addAll(rawTemplateLiteral.insertions);
+        insertionsList.expectedType = insertionsSym.type;
+
+        // Create an instance of the generated object class
+        BLangTypeInit typeNewExpr = ASTBuilderUtil.createEmptyTypeInit(pos, classObjType);
+        typeNewExpr.argsExpr.add(insertionsList);
+        typeNewExpr.initInvocation.argExprs.add(insertionsList);
+        typeNewExpr.initInvocation.requiredArgs.add(insertionsList);
+
+        result = rewriteExpr(typeNewExpr);
+    }
+
+    /**
+     * This method desugars a raw template literal object class for the provided raw template object type as follows:
+     * A literal defined as 'object:RawTemplate rt = `Hello ${name}!`;
+     * is desugared to,
+     *      type $anonType$0 object {
+     *          public string[] strings = ["Hello ", "!"];
```

Comment:
Shouldn't this be `readonly & string[]`

Reply:
Also can't we have this as an actual object defined in lang.object
https://github.com/ballerina-platform/ballerina-spec/issues/442#issuecomment-619482206

Reply:
Yes, need to change it to `readonly & string[]`.

Reply:
> Also can't we have this as an actual object defined in lang.object
> [ballerina-platform/ballerina-spec#442 (comment)](https://github.com/ballerina-platform/ballerina-spec/issues/442#issuecomment-619482206)

This is what I initially did. But as James clarified in the above issue, raw template literals are meant to work with abstract objects. If we try to use them with concrete objects, we run in to the issue I explained in https://github.com/ballerina-platform/ballerina-spec/issues/442#issuecomment-640633240.